### ISSUE TITLE
BUG without stack trace when public keys are not specified for an artifact cache using a self-signed certificate

### ISSUE THREAD

**Issue by Cynical-Optimist** [buildstream-migration/bst-staging#625]
[See original issue on GitLab](https://gitlab.com/BuildStream/buildstream/-/issues/625)
In GitLab by [[Gitlab user @ben-brewer-ct]](https://gitlab.com/ben-brewer-ct) on Sep 4, 2018, 15:21

## Summary

[//]: # (Summarize the bug encountered concisely)

## Steps to reproduce

This issue can be reproduced by configuring an artifact cache without providing the server-cert and client-cert fields.

```
artifacts:
  url: https://cache.trustable.io:11002
  client-key: certificates/client.key
  push: true
```

## What is the current bug behavior?

A bug is printed without any stack trace. The bug is not particularly meaningful.

```
[--:--:--][][] BUG     expected bytes, NoneType found

    TypeError: expected bytes, NoneType found
```

See full log below.

## What is the expected correct behavior?

It is expected that a readable error would be produced.
According to #buildstream it's also expected that a stack trace would be produced.

## Relevant logs and/or screenshots

https://gitlab.com/trustable/distros/minimal-distro/-/jobs/94525056

## Possible fixes

This isn't really a possible fix, but it'd be good if the documentation could be updated to emphasize the importance of providing public keys for self-signed certificates.

## Other relevant information

* BuildStream version affected: /milestone %BuildStream_v1.1.7

----

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Sep 4, 2018, 16:21

All configuration is supposed to be loaded by `_yaml.node_get()`, that causes the filename / line / column information to get printed in the error in the case an expected key is not found.

Since the artifact cache parses it's config in a bit of an unorthodox way, it's possible it forgot to use the appropriate `_yaml` accessor.

That would be the fix for the specific case, the fact that the BUG came without a stack trace is separate.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @toscalix]](https://gitlab.com/toscalix) on Sep 17, 2018, 09:27

assigned to [[Gitlab user @ben-brewer-ct]](https://gitlab.com/ben-brewer-ct)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @toscalix]](https://gitlab.com/toscalix) on Sep 17, 2018, 09:27

unassigned [[Gitlab user @ben-brewer-ct]](https://gitlab.com/ben-brewer-ct)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Sep 18, 2018, 10:18

mentioned in commit d250a6cd4a6f2a4493b95f8228e00cf2b26dfb1a

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Sep 18, 2018, 10:21

mentioned in commit acd11d0774fcf1c357cff51de277e2333b3827c5

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Sep 18, 2018, 10:38

mentioned in commit 0c426cb43b43305359150654917495a87a2a3ce4

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Sep 18, 2018, 10:38

mentioned in commit 8f4aa67e09ff690c43ceef5a5f7a08bc2bd3d09a

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Sep 18, 2018, 10:56

mentioned in commit cdd9a58678b54155c47821899fd3ccf89b895280

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Sep 18, 2018, 10:56

mentioned in commit 30b41959646e9076f45e21089216a27c1088cdb1

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Sep 18, 2018, 11:16

closed via merge request !804

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Sep 18, 2018, 11:16

mentioned in commit 97071b6ee598aad5ed159bb038bb333a468f065e

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Sep 18, 2018, 11:17

This is fixed with !804 and !805 (for `bst-1.2`).

The fix will be released with `1.2.1` which is planned by the end of this week.
